src/control/system/raft/database_members.go

```diff
@@ -47,28 +48,62 @@ func (mrm MemberRankMap) MarshalJSON() ([]byte, error) {
 	return json.Marshal(jm)
 }
 
-func (mam MemberAddrMap) addMember(addr *net.TCPAddr, m *system.Member) {
-	if _, exists := mam[addr.String()]; !exists {
-		mam[addr.String()] = []*system.Member{}
+// dedupeMembers removes members with duplicate UUIDs from MemberAddrMap.
+func (mam MemberAddrMap) dedupeMembers(addrStr string) {
+	seen := make(map[string]struct{})
+
+	newMembers := system.Members{}
+	for _, m := range mam[addrStr] {
+		us := m.UUID.String()
+
+		if _, exists := seen[us]; !exists {
+			seen[us] = struct{}{}
+			newMembers = append(newMembers, m)
+		}
 	}
-	mam[addr.String()] = append(mam[addr.String()], m)
+
+	mam[addrStr] = newMembers
+}
+
+func (mam MemberAddrMap) addMember(addr *net.TCPAddr, m *system.Member) {
+	as := addr.String()
+
+	mam[as] = append(mam[as], m)
+	mam.dedupeMembers(as)
```

Comment:
I'm going to reiterate my comments from standup, here. Succinctly, it seems wrong to me that this dedupe functionality should be called as part of regular membership management rather than fixing the broken code that allows the duplicates to be introduced in the first place. This code addresses a specific symptom seen in this scenario, but there may be other symptoms that have not yet been encountered. Without understanding and addressing the root of the problem, how can you have confidence that there won't be other production outages?

I do think it would be useful to keep the dedupe and other repair functionality developed to address the bug as part of a CR command, though.

Reply:
The dedupe functionality covers an observed problem that the DAOS raft library wrappers exhibit. The symptom is well defined and the join-replace functionality relatively simple. I have manually verified the fix extensively with @knard38 and observed that we can resolve the inaccuracies in the generated snapshot using a small update to the wrapping logic. I believe the issue is resolved by the update to removeMember so I'm attempting to verify the fix without the dedupe function. Both customers are eagerly awaiting the replace feature. The fix should be landed to 2.6.4 ASAP. @kjacque has also written a unit test which I will integrate into the PR, thanks!